File: mocks/oppgave-mock/src/main/java/no/nav/pensjon/vtp/mocks/oppgave/rest/infrastruktur/validering/CountFieldsMatching.java
package no.nav.pensjon.vtp.mocks.oppgave.rest.infrastruktur.validering;

import static java.util.Arrays.stream;

import static org.springframework.beans.PropertyAccessorFactory.forBeanPropertyAccess;

import org.springframework.beans.BeanWrapper;

class CountFieldsMatching {
    public static Long count(Object o, String[] fields) {
        final BeanWrapper beanWrapper = forBeanPropertyAccess(o);

        return stream(fields)
                .map(beanWrapper::getPropertyValue)
                .filter(CountFieldsMatching::isNotBlank)
                .count();
    }

    static boolean isNotBlank(final Object object) {
        return !isBlank(object);
    }

    static boolean isBlank(final Object object) {
        return object == null || object.toString().chars().allMatch(Character::isWhitespace);
    }
}

File: vtp-pensjon-application/src/main/kotlin/no/nav/pensjon/vtp/mocks/oppgave/rest/infrastruktur/validering/CountFieldsMatching.kt
package no.nav.pensjon.vtp.mocks.oppgave.rest.infrastruktur.validering

import org.springframework.beans.PropertyAccessorFactory

fun countFieldsMatching(any: Any, fields: Array<String>): Int {
    val beanWrapper = PropertyAccessorFactory.forBeanPropertyAccess(any)

    return fields.toList()
        .map { beanWrapper.getPropertyValue(it) }
        .filter { isNotBlank(it) }
        .count()
}

private fun isNotBlank(any: Any?): Boolean {
    return !isBlank(any)
}

private fun isBlank(any: Any?): Boolean {
    return any == null || any.toString().chars()
        .allMatch { codePoint: Int -> Character.isWhitespace(codePoint) }
}
